Create a DDPLoggingData and expose it to python interface

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50637 Create PyTorch DDP logging APIs for applications to use
* **#50622 Create a DDPLoggingData and expose it to python interface**

1. Define a DDPLoggingData struct that is the placeholder for all the ddp related logging fields
2. Put the DDPLoggingData struct in the C10 directory so that it can be easily imported by c10 and torch files
3. Expose get_ddp_logging_data() method in python so that users can get the logging data and dump in their applications
4. Unit test tested the logging data can be set and got as expected
5. Follow up will add more logging fields such as perf stats, internal states, env variables and etc

Differential Revision: [D25930527](https://our.internmc.facebook.com/intern/diff/D25930527/)

zhaojuanmao committed:
Create a DdpLogger and expose it to python interface

1. Define a DDPLoggingData struct that is the placeholder for all the ddp related logging fields
2. Put the DDPLoggingData struct in the C10 directory so that it can be easily imported by c10 and torch files
3. Expose get_ddp_logging_data() method in python so that users can get the logging data and dump in their applications
4. Unit test tested the logging data can be set and got as expected
5. Follow up will add more logging fields such as perf stats, internal states, env variables and etc

Differential Revision: [D25930527](https://our.internmc.facebook.com/intern/diff/D25930527/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -105,6 +105,19 @@ class Reducer {
 
+  // Set logging data that can be got during DistributedDataParallel
+  // construction time.
+  void set_construction_logging_data(
```
In order to avoid duplicates/identical data should we log only if rank == 0?

zhaojuanmao replied:
i think it is better to log all ranks' data, especially if we want to debug desync issue, all ranks' data will be useful

rohan-varma commented on c10/util/Logging.h:
```diff
@@ -300,6 +300,29 @@ BINARY_COMP_HELPER(LessEquals, <=)
+// TODO -- right now starting with logging a small set of straightforward
+// fields, will add more fields as follow ups such as performance stats,
+// internal states and env variables and etc.
+struct DDPLoggingData {
```
General question - it looks like all this data is logged once per DDP construction. Can it be easily extended to support logging fields that are computed/can change at runtime during DDP training?

zhaojuanmao replied:
yes, logging fields can be filled during runtime as well and logged at runtime

SciPioneer commented on c10/util/Logging.h:
```diff
@@ -300,6 +300,29 @@ BINARY_COMP_HELPER(LessEquals, <=)
+// fields, will add more fields as follow ups such as performance stats,
+// internal states and env variables and etc.
+struct DDPLoggingData {
+ public:
```
Nit: No need to specify "public" is in a struct.

zhaojuanmao replied:
oops, forgot to remove it when I changed it from class to struct

SciPioneer commented on c10/util/Logging.h:
```diff
@@ -300,6 +300,29 @@ BINARY_COMP_HELPER(LessEquals, <=)
+// TODO -- right now starting with logging a small set of straightforward
+// fields, will add more fields as follow ups such as performance stats,
+// internal states and env variables and etc.
+struct DDPLoggingData {
```
[Optional] Your design has a few logging data categories. I would suggest that we should map each data category to a separate struct inside this one. Grouping relevant fields will make this struct more manageable. Since this is the starting point, it is also fine to keep the structure flat for now.

zhaojuanmao replied:
sounds good! let me organize the fields into separate struct in followup diff

SciPioneer commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1465,6 +1466,29 @@ void Reducer::ensure_prior_reduction_finished() {
+  const std::string& backend,
+  const std::string& module_name,
+  const std::vector<int>& device_ids,
+  int output_device,
```
In the unit test, output_device is expected to be -1. Probably add some comments to explain when should -1 be used?
Otherwise, I have to look into the implementation details (`dist._set_construction_logging_data` in distributed.py)

zhaojuanmao replied:
sounds good!

zhaojuanmao committed:
Update on "Create a DDPLoggingData and expose it to python interface"


1. Define a DDPLoggingData struct that is the placeholder for all the ddp related logging fields
2. Put the DDPLoggingData struct in the C10 directory so that it can be easily imported by c10 and torch files
3. Expose get_ddp_logging_data() method in python so that users can get the logging data and dump in their applications
4. Unit test tested the logging data can be set and got as expected
5. Follow up will add more logging fields such as perf stats, internal states, env variables and etc

Differential Revision: [D25930527](https://our.internmc.facebook.com/intern/diff/D25930527/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Create a DDPLoggingData and expose it to python interface"


1. Define a DDPLoggingData struct that is the placeholder for all the ddp related logging fields
2. Put the DDPLoggingData struct in the C10 directory so that it can be easily imported by c10 and torch files
3. Expose get_ddp_logging_data() method in python so that users can get the logging data and dump in their applications
4. Unit test tested the logging data can be set and got as expected
5. Follow up will add more logging fields such as perf stats, internal states, env variables and etc

Differential Revision: [D25930527](https://our.internmc.facebook.com/intern/diff/D25930527/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Create a DDPLoggingData and expose it to python interface"


1. Define a DDPLoggingData struct that is the placeholder for all the ddp related logging fields
2. Put the DDPLoggingData struct in the C10 directory so that it can be easily imported by c10 and torch files
3. Expose get_ddp_logging_data() method in python so that users can get the logging data and dump in their applications
4. Unit test tested the logging data can be set and got as expected
5. Follow up will add more logging fields such as perf stats, internal states, env variables and etc

Differential Revision: [D25930527](https://our.internmc.facebook.com/intern/diff/D25930527/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Create a DDPLoggingData and expose it to python interface"


1. Define a DDPLoggingData struct that is the placeholder for all the ddp related logging fields
2. Put the DDPLoggingData struct in the C10 directory so that it can be easily imported by c10 and torch files
3. Expose get_ddp_logging_data() method in python so that users can get the logging data and dump in their applications
4. Unit test tested the logging data can be set and got as expected
5. Follow up will add more logging fields such as perf stats, internal states, env variables and etc

Differential Revision: [D25930527](https://our.internmc.facebook.com/intern/diff/D25930527/)

[ghstack-poisoned]
